mlflow/recipes/steps/train.py

```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
                 processed_data[key] = float(value)
             elif isinstance(value, np.integer):
                 processed_data[key] = int(value)
+            elif isinstance(value, dict):
+                processed_data[key] = str(value)
             else:
                 processed_data[key] = value
 
         if len(processed_data) > 0:
             yaml.safe_dump(processed_data, file, **kwargs)
+
+    def _rebalance_classes(self, train_df):
+        import pandas as pd
+        import numpy as np
+        from sklearn.utils.class_weight import compute_class_weight
+
+        resampling_minority_percentage = self.step_config.get(
+            "resampling_minority_percentage", _REBALANCING_DEFAULT_RATIO
+        )
+
+        df_positive_class = train_df[train_df[self.target_col] == self.positive_class]
+        df_negative_class = train_df[train_df[self.target_col] != self.positive_class]
+
+        classes = np.unique(train_df[self.target_col])
+        class_weights = compute_class_weight(
+            class_weight="balanced",
+            classes=classes,
+            y=train_df[self.target_col],
+        )
+
+        self.original_class_weights = dict(zip(classes, class_weights))
+
+        if len(df_positive_class) > len(df_negative_class):
+            df_minority_class, df_majority_class = df_negative_class, df_positive_class
+            original_minority_percentage = len(df_negative_class) / len(train_df)
+        else:
+            df_minority_class, df_majority_class = df_positive_class, df_negative_class
+            original_minority_percentage = len(df_positive_class) / len(train_df)
+
+        if original_minority_percentage >= resampling_minority_percentage:
+            _logger.info(
+                f"Class imbalance of {original_minority_percentage:.2f} "
+                f"is better than {resampling_minority_percentage}, no need to rebalance"
+            )
+            return train_df
+
+        _logger.info(
+            f"Detected class imbalance: original count - minority class: "
+            f"{len(df_minority_class)} vs majority class: {len(df_majority_class)}"
+        )
+
+        majority_class_target = int(
+            len(df_minority_class)
+            * (1 - resampling_minority_percentage)
+            / resampling_minority_percentage
+        )
+        df_majority_downsampled = df_majority_class.sample(majority_class_target)
+        _logger.info(
+            f"After downsampling: original count - minority class: "
+            f"{len(df_minority_class)} vs majority class: {len(df_majority_downsampled)}"
+        )
+        train_df = pd.concat([df_minority_class, df_majority_downsampled], axis=0)
```

Comment:
This is dangerous. Should `shuffle` to make sure the SGD optimizer does't go crazy over one class for long period of time:)